// app/block/_client/views/widgets/Area.js
import React from "react"
import {findDOMNode} from "react-dom"
import {defaultsDeep, filter} from "app/_shared/utils"
import {getDefaultOptions, drawChart} from "app/block/_client/views/widgets/chart"

const {Component, PropTypes} = React

class Area extends Component {

  static propTypes = {
    widget: PropTypes.object.isRequired,
    block: PropTypes.object.isRequired,
    data: PropTypes.array.isRequired
  }

  componentDidMount() {
    this.drawChart()
  }

  drawChart() {
    const {widget, block, data} = this.props
    const chartOptions = defaultsDeep({
      chart: {
        type: "area"
      }
    }, getDefaultOptions(widget, block))
    if (widget.stacking) {
      chartOptions.plotOptions.series.stacking = widget.stacking
    }
    chartOptions.xAxis.categories = data.map(record => record.time)
    chartOptions.series = block.data.category.map(category => ({
      name: category.id,
      data: filter(data, record => record.category === category.id)
        .map(record => +record.value.replace(/,/g, ""))
    }))
    drawChart(findDOMNode(this), chartOptions)
  }

  render() {
    return <div></div>
  }

}

export default Area
