// app/block/_client/views/widgets/Bar.js
import React from "react"
import {findDOMNode} from "react-dom"
import {defaultsDeep, isString} from "app/_shared/utils"
import {getDefaultOptions, drawChart} from "app/block/_client/views/widgets/chart"

const {Component, PropTypes} = React

class Bar extends Component {

  static propTypes = {
    widget: PropTypes.object.isRequired,
    block: PropTypes.object.isRequired,
    data: PropTypes.array.isRequired
  }

  componentDidMount() {
    this.drawChart()
  }

  drawChart() {
    const {widget, block, data} = this.props
    const chartOptions = defaultsDeep({
      chart: {
        type: "column"
      }
    }, getDefaultOptions(widget, block))
    if (widget.categories) {
      chartOptions.xAxis.categories = widget.series.map(series => series.label)
      chartOptions.plotOptions.column = {
        grouping: false
      }
    } else {
      chartOptions.xAxis.categories = data.map(record => record.time)
    }
    if (widget.stacking) {
      chartOptions.plotOptions.series.stacking = widget.stacking
    }
    chartOptions.series = widget.series.map(series => ({
      name: series.label,
      data: data.map(record => isString(record[series.id]) ? +record[series.id].replace(/,/g, "") : null)
    }))
    drawChart(findDOMNode(this), chartOptions)
  }

  render() {
    return <div></div>
  }

}

export default Bar
