// app/block/_client/views/widgets/Bar.js
import React from "react"
import {findDOMNode} from "react-dom"

const {Component, PropTypes} = React

class Bar extends Component {

  static propTypes = {
    widget: PropTypes.object.isRequired,
    block: PropTypes.object.isRequired,
    data: PropTypes.array.isRequired
  }

  componentDidMount() {
    this.drawChart()
  }

  drawChart() {
    const {widget, data} = this.props
    const chartOptions = {
      title: false,
      credits: false,
      chart: {
        type: "column"
      }
    }
    const categories = data.map(record => record[widget.category])
    chartOptions.xAxis = {
      title: {
        text: widget.xAxis.label
      },
      categories
    }
    chartOptions.yAxis = {
      title: {
        text: widget.yAxis.label
      }
    }
    chartOptions.series = widget.series.map(series => ({
      name: series.label,
      data: data.map(record => +record[series.id].replace(/,/g, ""))
    }))
    const {$} = window
    $(findDOMNode(this)).highcharts(chartOptions)
  }

  render() {
    return <div></div>
  }

}

export default Bar
