// app/block/_client/views/widgets/Pie.js
import React from "react"
import {findDOMNode} from "react-dom"
import {defaultsDeep, find} from "app/_shared/utils"
import {getDefaultOptions, drawChart} from "app/block/_client/views/widgets/chart"

const {Component, PropTypes} = React

class Pie extends Component {

  static propTypes = {
    widget: PropTypes.object.isRequired,
    block: PropTypes.object.isRequired,
    data: PropTypes.array.isRequired
  }

  componentDidMount() {
    this.drawChart()
  }

  drawChart() {
    const {widget, block, data} = this.props
    const chartOptions = defaultsDeep({
      chart: {
        type: "pie"
      }
    }, getDefaultOptions(widget, block))
    chartOptions.xAxis.categories = block.data.category.map(category => category.id)
    chartOptions.plotOptions.pie = {
      dataLabels: false,
      showInLegend: true
    }
    chartOptions.series = [{
      colorByPoint: true,
      data: block.data.category.map(category => ({
        name: category.id,
        y: +find(data, record => record.category === category.id).value.replace(/,/g, "")
      }))
    }]
    drawChart(findDOMNode(this), chartOptions)
  }

  render() {
    return <div></div>
  }

}

export default Pie
