// app/block/_client/views/widgets/Pie.js
import React from "react"
import {findDOMNode} from "react-dom"
import {defaultsDeep, find, isString} from "app/_shared/utils"
import {getDefaultOptions, drawChart} from "app/block/_client/views/widgets/chart"

const {Component, PropTypes} = React

class Pie extends Component {

  static propTypes = {
    widget: PropTypes.object.isRequired,
    block: PropTypes.object.isRequired,
    data: PropTypes.array.isRequired
  }

  componentDidMount() {
    this.drawChart()
  }

  drawChart() {
    const {widget, block, data} = this.props
    const chartOptions = defaultsDeep({
      chart: {
        type: "pie"
      }
    }, getDefaultOptions(widget, block))
    chartOptions.xAxis.categories = widget.series.map(series => series.label)
    chartOptions.plotOptions.pie = {
      dataLabels: false,
      showInLegend: true
    }
    chartOptions.series = [{
      name: widget.yAxis.label,
      colorByPoint: true,
      data: widget.series.map(series => ({
        name: series.label,
        y: +find(data, record => isString(record[series.id]))[series.id].replace(/,/g, "")
      }))
    }]
    drawChart(findDOMNode(this), chartOptions)
  }

  render() {
    return <div></div>
  }

}

export default Pie
