// app/web/actions/form.js
import {createAction} from "redux-actions"
import {post} from "app/core/utils/ajax"

export const INIT_FORM = "INIT_FORM"
export const REMOVE_FORM = "REMOVE_FORM"
export const REQUEST_RESULT = "REQUEST_RESULT"
export const RECEIVE_RESULT = "RECEIVE_RESULT"
export const REJECT_RESULT = "REJECT_RESULT"
export const SET_DATA_SOURCE = "SET_DATA_SOURCE"

export const initForm = createAction(INIT_FORM)
export const removeForm = createAction(REMOVE_FORM)
export const requestResult = createAction(REQUEST_RESULT)
export const receiveResult = createAction(RECEIVE_RESULT)
export const rejectResult = createAction(REJECT_RESULT)

export const setDataSource = createAction(SET_DATA_SOURCE, (id, name, data) => ({
  id,
  name,
  data
}))

export function submitForm(id, data, done) {
  return (dispatch, getState) => {
    const {info, form: {[id]: form}} = getState()
    dispatch(requestResult({
      id,
      data
    }))
    return post(`${info.apiUrl}${form.url}`, {}, data).then(result => {
      dispatch(receiveResult({
        id,
        result
      }))
      done(result)
    })
    .catch(error => {
      dispatch(rejectResult({
        id,
        error: error.message || error.error
      }))
    })
  }
}
