// app/misc/cdn.js
import fs from "fs"
import path from "path"
import Download from "download"
import config from "app/config"

function getLibName(name) {
  return name.split("/")[0]
}

export function register(server, options, next) {
  if (config.cdn.cache) {
    const cdnPath = path.resolve("tmp/cdn")
    const cdnUrlByLibName = {}
    config.cdn.libs.forEach(lib => {
      lib.styles.forEach(style => {
        cdnUrlByLibName[getLibName(style)] = lib.url
      })
      lib.scripts.forEach(script => {
        cdnUrlByLibName[getLibName(script)] = lib.url
      })
    })
    server.route({
      method: "GET",
      path: `${config.cdn.cache}/{path*}`,
      handler: (request, reply) => {
        const filePath = path.join(cdnPath, request.params.path)
        fs.stat(filePath, notFound => {
          if (notFound) {
            new Download()
              .get(`${cdnUrlByLibName[getLibName(request.params.path)]}/${request.params.path}`)
              .dest(path.join(filePath, ".."))
              .run(() => {
                reply.file(filePath)
              })
            return
          }
          reply.file(filePath)
        })
      }
    })
  }
  next()
}

register.attributes = {
  name: "agridash-vendors"
}
