// app/report/_client/views/WidgetChartLine.js
import React from "react"
import {findDOMNode} from "react-dom"

const {Component, PropTypes} = React
const {$} = window

class WidgetChartLine extends Component {

  static propTypes = {
    block: PropTypes.object.isRequired,
    config: PropTypes.object.isRequired,
    data: PropTypes.array.isRequired
  }

  componentDidMount() {
    this.updateChart()
  }

  updateChart() {
    const {block, config, data} = this.props
    const chartOptions = {
      title: false
    }
    const categories = []
    for (let year = block.loader.year[0]; year <= block.loader.year[1]; ++year) {
      categories.push(year)
    }
    chartOptions.xAxis = {
      categories
    }
    chartOptions.yAxis = {
      title: {
        text: config.yAxis.label
      }
    }
    chartOptions.legend = {
      layout: "vertical",
      align: "right",
      verticalAlign: "middle",
      borderWidth: 0
    }
    const allSeries = []
    config.series.forEach(seriesConfig => {
      const series = {
        name: seriesConfig.label,
        data: data.map(record => +record[seriesConfig.id].replace(/,/g, ""))
      }
      allSeries.push(series)
    })
    chartOptions.series = allSeries
    $(findDOMNode(this)).highcharts(chartOptions)
  }

  render() {
    return <div></div>
  }

}

export default WidgetChartLine
