// app/web/client/views/Map.js
import React from "react"
import {findDOMNode} from "react-dom"
import styles from "app/web/client/views/Map.css"

const {Component} = React

class Map extends Component {

  componentDidMount() {
    const {L, clientConfig} = window
    const map = L.map(findDOMNode(this), {
      attributionControl: false,
      zoomAnimation: false,
      center: [37.8, -96.9],
      zoom: 4
    })
    // L.tileLayer("https://{s}.tiles.mapbox.com/v4/{style}/{z}/{x}/{y}.png?access_token={key}", {
    //   style: "mapbox.light",
    //   key: "pk.eyJ1IjoiYWdyaWRhc2giLCJhIjoiYjdiZjc4YTA0ZjhhODg0N2VkMDM0ODQwNGUwNDA0ZDQifQ.jmfdrMQc252iW2ovBPRTOg"
    // }).addTo(map)
    L.tileLayer(`${clientConfig.tileUrl}/area/{z}/{x}/{y}.png?s=US_COUNTY`).addTo(map)
    L.tileLayer(`${clientConfig.tileUrl}/heat/{z}/{x}/{y}.png?s=US_COUNTY`).addTo(map)
    L.tileLayer(`${clientConfig.tileUrl}/bubble/{z}/{x}/{y}.png?s=US_STATE`).addTo(map)
    L.tileLayer(`${clientConfig.tileUrl}/dot/{z}/{x}/{y}.png?s=US_COUNTY`).addTo(map)
    L.tileLayer(`${clientConfig.tileUrl}/column/{z}/{x}/{y}.png?s=US_STATE`).addTo(map)
  }

  render() {
    return <div className={styles.container}>
    </div>
  }

}

export default Map
