// app/core/utils/cache.js
import {createClient} from "redis"
import {createLogger} from "bunyan"
import {isUndefined, isNull} from "app/core/utils"
import {redis as redisConfig} from "app/config"

export const TTL_XSHORT = 5 * 60
export const TTL_SHORT = 15 * 60
export const TTL_LONG = 60 * 60

const log = createLogger({name: "cache"})

const redis = createClient(redisConfig.port || 6379, redisConfig.host, {
  "auth_pass": redisConfig.password
})

redis.on("error", error => {
  log.error(error)
})

redis.select(redisConfig.database || 0, error => {
  if (error) {
    log.error(error)
    return
  }
  log.info({
    cache: `${redis.address}#${redis.selected_db}`
  })
})

function promisify(command) {
  return function(...args) {
    return new Promise((resolve, reject) => {
      redis[command](...args, (error, response) => {
        if (error) {
          reject(error)
          return
        }
        resolve(response)
      })
    })
  }
}

export const set = promisify("set")
export const setex = promisify("setex")
export const get = promisify("get")

export function fromCache(key, time, request) {
  return get(key).then(response => {
    if (response) {
      return JSON.parse(response).value
    }
    return Promise.resolve(request(key)).then(value => {
      if (time) {
        const timeToCache = (isUndefined(value) || isNull(value)) ? TTL_XSHORT : time
        return setex(key, timeToCache, JSON.stringify({value})).then(() => value)
      }
      return set(key, JSON.stringify({value})).then(() => value)
    })
  })
}
