// app/web/forms/category.js
import {getSchema as getIdSchema, getOptions as getIdOptions} from "app/web/forms/id"
import {getSchema as getNameSchema, getOptions as getNameOptions} from "app/web/forms/name"
import {getSchema as getCategorySourceSchema, getOptions as getCategorySourceOptions}
  from "app/web/forms/categorySource"
import {getSchema as getMockCategorySchema, getOptions as getMockCategoryOptions} from "app/web/forms/mockCategory"

export function getSchema() {
  return {
    type: "object",
    properties: {
      id: getIdSchema(),
      name: getNameSchema(),
      source: getCategorySourceSchema(),
      mockCategory: getMockCategorySchema()
    },
    dependencies: {
      mockCategory: "source"
    }
  }
}

export function getOptions() {
  return {
    label: "Category",
    fields: {
      id: getIdOptions(),
      name: getNameOptions(),
      source: getCategorySourceOptions(),
      mockCategory: getMockCategoryOptions({
        dependencies: {
          source: "mock"
        }
      })
    }
  }
}
