// app/web/forms/createFormOptions.js
function createFormOptions(submitLabel, submitCallback, options) {
  const {$} = window
  return {
    toggleSubmitValidState: false,
    buttons: {
      submit: {
        title: submitLabel,
        styles: "btn btn-primary",
        click(event) {
          this.refreshValidationState(true)
          const buttonsContainer = $(event.target).parent()
          if (this.isValid()) {
            if (buttonsContainer.children().length !== 1) {
              buttonsContainer.children().first().remove()
            }
            submitCallback(this.getValue())
          } else if (buttonsContainer.children().length === 1) {
            buttonsContainer.prepend($("<span></span>")
              .html("The form is not valid.")
              .attr("class", "label label-danger")
              .css({
                marginRight: "1rem",
                fontSize: "110%"
              }))
            setTimeout(() => {
              if (buttonsContainer.children().length !== 1) {
                buttonsContainer.children().first().remove()
              }
            }, 5000)
          }
        }
      }
    },
    ...options
  }
}

export default createFormOptions
