// app/web/forms/layer.js
import createSelectOptions from "app/web/forms/createSelectOptions"
import {getSchema as getLocationTypeSchema, getOptions as getLocationTypeOptions} from "app/web/forms/locationType"

export function getSchema() {
  return {
    type: "object",
    properties: {
      type: {
        type: "string",
        required: true
      },
      locationType: getLocationTypeSchema()
    }
  }
}

export function getOptions() {
  return {
    label: "Layer",
    fields: {
      type: createSelectOptions({
        label: "Type",
        dataSource: [
          {
            value: "area",
            text: "Area"
          },
          {
            value: "heat",
            text: "Heat"
          },
          {
            value: "dot",
            text: "Dot"
          },
          {
            value: "bubble",
            text: "Bubble"
          },
          {
            value: "column",
            text: "Column"
          }
        ]
      }),
      locationType: getLocationTypeOptions()
    }
  }
}
