// app/web/forms/locationType.js
import store from "app/web/store"
import {fetchGeos} from "app/web/actions/geo"
import createSelectOptions from "app/web/forms/createSelectOptions"

export function getSchema() {
  return {
    type: "string",
    required: true,
    default: "US_STATE"
  }
}

export function getOptions(options) {
  return createSelectOptions({
    label: "Location Type",
    dataSource: [
      {
        value: "US_STATE",
        text: "States"
      },
      {
        value: "US_COUNTY",
        text: "Counties"
      }
    ],
    ...options
  })
}

export function updateLocations(field, locationsField) {
  const locationType = field.getValue()
  if (locationType) {
    locationsField.showLoading()
    store.dispatch(fetchGeos(locationType)).then(() => {
      const {geo: geoState} = store.getState()
      locationsField.options.dataSource = geoState[locationType].geos.map(geo => ({
        value: geo.code,
        text: geo.name
      }))
      locationsField.refresh()
      locationsField.hideLoading()
    })
  }
}
