// app/web/forms/locationType.js
import {dispatch, getState} from "app/web/store"
import {getGeos} from "app/web/actions/model"
import {getSelectOptions} from "app/web/forms/common"

export function getSchema() {
  return {
    type: "string",
    required: true,
    default: "US_STATE"
  }
}

export function getOptions(options) {
  return getSelectOptions({
    label: "Location Type",
    dataSource: [
      {
        value: "US_STATE",
        text: "States"
      },
      {
        value: "US_COUNTY",
        text: "Counties"
      }
    ],
    ...options
  })
}

export function updateLocations(field, locationsField) {
  const locationType = field.getValue()
  if (locationType) {
    locationsField.showLoading()
    dispatch(getGeos(locationType))
    .then(() => {
      const {model: {geosBySource: {[locationType]: geos}}} = getState()
      locationsField.options.dataSource = geos.map(geo => ({
        value: geo.code,
        text: geo.name
      }))
      locationsField.refresh()
      locationsField.hideLoading()
    })
  }
}
