// app/web/forms/report.js
import {createArrayOptions} from "app/web/forms/common"
import {getSchema as getNameSchema, getOptions as getNameOptions} from "app/web/forms/name"
import {getSchema as getNextIdSchema, getOptions as getNextIdOptions, updateId} from "app/web/forms/nextId"
import {getSchema as getCategorySchema, getOptions as getCategoryOptions} from "app/web/forms/category"
import {getSchema as getViewSchema, getOptions as getViewOptions} from "app/web/forms/view"

export function getSchema() {
  return {
    type: "object",
    properties: {
      name: getNameSchema(),
      nextCategoryId: getNextIdSchema(),
      nextViewId: getNextIdSchema(),
      categories: {
        type: "array",
        items: getCategorySchema()
      },
      views: {
        type: "array",
        items: getViewSchema()
      }
    }
  }
}

export function getOptions(options) {
  return {
    label: "Report",
    fields: {
      name: getNameOptions(),
      nextCategoryId: getNextIdOptions(),
      nextViewId: getNextIdOptions(),
      categories: createArrayOptions({
        label: "Categories",
        items: getCategoryOptions(),
        postRender(done) {
          this.on("add", categoryField => {
            updateId(this.parent, categoryField, this.parent.childrenByPropertyId.nextCategoryId)
          })
          done()
        }
      }),
      views: createArrayOptions({
        label: "Views",
        items: getViewOptions(),
        postRender(done) {
          this.on("add", viewField => {
            updateId(this.parent, viewField, this.parent.childrenByPropertyId.nextViewId)
          })
          done()
        }
      })
    },
    ...options
  }
}
