// app/web/forms/report.js
import createArrayOptions from "app/web/forms/createArrayOptions"
import {getSchema as getViewSchema, getOptions as getViewOptions} from "app/web/forms/view"

export function getSchema() {
  return {
    type: "object",
    properties: {
      name: {
        type: "string",
        required: true
      },
      nextViewId: {
        type: "number",
        default: 1,
        required: true
      },
      views: {
        type: "array",
        items: getViewSchema()
      }
    }
  }
}

function updateViewId(field, viewField) {
  const {nextViewId: nextViewIdField} = field.childrenByPropertyId
  const id = nextViewIdField.getValue()
  const {id: viewIdField} = viewField.childrenByPropertyId
  viewIdField.setValue(id)
  nextViewIdField.setValue(id + 1)
}

export function getOptions(options) {
  return {
    label: "Report",
    fields: {
      name: {
        label: "Name"
      },
      nextViewId: {
        type: "hidden"
      },
      views: createArrayOptions({
        label: "Views",
        items: getViewOptions(),
        postRender(done) {
          this.on("add", viewField => {
            updateViewId(this.parent, viewField)
          })
          done()
        }
      })
    },
    ...options
  }
}
