// app/core/models/user.js
import sql from "app/core/utils/sql"
import {fromCache, del, TTL_SHORT, TTL_LONG} from "app/core/utils/cache"

function beforeCache(user) {
  if (!user) {
    return user
  }
  const {createdat: createdAt, ...others} = user
  return {
    ...others,
    createdAt
  }
}

function afterCache(user) {
  if (!user) {
    return user
  }
  const {hash, ...userWithoutHash} = user
  return userWithoutHash
}

export function findById(id) {
  return fromCache(`user/${id}`, TTL_LONG, () => sql.first().from("appuser").where("id", id), beforeCache, afterCache)
}

export function findByUsername(username, requireHash) {
  return fromCache(`username/${username}`, TTL_SHORT, () => sql.first().from("appuser")
  .where("username", username), beforeCache, user => {
    if (requireHash) {
      return user
    }
    return afterCache(user)
  })
}

export function clearCache(id, username) {
  return del(`user/${id}`, `username/${username}`)
}

export function create(user) {
  return sql.insert(user).into("appuser").returning("id")
  .then(([id]) => {
    return clearCache(id, user.username)
    .then(() => findById(id))
  })
}
