// app/_shared/models/block.js
import {find} from "app/_shared/utils"

const categories = [
  {
    id: 1,
    type: "mock"
  },
  {
    id: 2,
    type: "mock"
  },
  {
    id: 3,
    type: "mock"
  }
]

const blocks = [
  {
    id: 1,
    title: "Bar: compare different categories",
    data: {
      category: categories,
      time: 2014
    },
    widget: {
      type: "bar",
      categories: true
    }
  },
  {
    id: 2,
    title: "Bar: show changes over time",
    data: {
      category: categories,
      time: [2010, 2014]
    },
    widget: {
      type: "bar"
    }
  },
  {
    id: 3,
    title: "Bar (stacked): compare parts of a whole",
    data: {
      category: categories,
      time: [2010, 2014]
    },
    widget: {
      type: "bar",
      stacking: "normal"
    }
  },
  {
    id: 4,
    title: "Pie: compare parts of a whole",
    data: {
      category: categories,
      time: 2014
    },
    widget: {
      type: "pie"
    }
  },
  {
    id: 5,
    title: "Line: show changes over time",
    data: {
      category: categories,
      time: [2000, 2014]
    },
    widget: {
      type: "line"
    }
  },
  {
    id: 6,
    title: "Area: show changes over time",
    data: {
      category: categories,
      time: [2000, 2014]
    },
    widget: {
      type: "area"
    }
  },
  {
    id: 7,
    title: "Area (100% stacked): show changes over time",
    data: {
      category: categories,
      time: [2000, 2014]
    },
    widget: {
      type: "area",
      stacking: "percent"
    }
  }
]

export function findAll() {
  return Promise.resolve(blocks)
}

export function findById(id) {
  return Promise.resolve(find(blocks, block => block.id === id))
}
