// app/_shared/models/block.js
import {find} from "app/_shared/utils"

const timeRange = [2010, 2014]

const categories = [
  {
    id: "a"
  },
  {
    id: "b"
  },
  {
    id: "c"
  }
]

const series = [
  {
    id: "a",
    label: "A"
  },
  {
    id: "b",
    label: "B"
  },
  {
    id: "c",
    label: "C"
  }
]

const blocks = [
  {
    id: 1,
    title: "Bar: compare different categories",
    loader: {
      type: "mock",
      categories
    },
    widget: {
      type: "bar",
      categories: true,
      series,
      yAxis: {
        label: "Value"
      }
    }
  },
  {
    id: 2,
    title: "Bar: show change over time",
    loader: {
      type: "mock",
      categories,
      timeRange
    },
    widget: {
      type: "bar",
      series
    }
  },
  {
    id: 3,
    title: "Bar (stacked): compare parts of a whole",
    loader: {
      type: "mock",
      categories,
      timeRange
    },
    widget: {
      type: "bar",
      stacking: "normal",
      series
    }
  },
  {
    id: 4,
    title: "Bar (100% stacked): compare parts of a whole",
    loader: {
      type: "mock",
      categories,
      timeRange
    },
    widget: {
      type: "bar",
      stacking: "percent",
      series
    }
  },
  {
    id: 5,
    title: "Pie: compare parts of a whole",
    loader: {
      type: "mock",
      categories
    },
    widget: {
      type: "pie",
      series,
      yAxis: {
        label: "Value"
      }
    }
  }
]

export function findAll() {
  return Promise.resolve(blocks)
}

export function findById(id) {
  return Promise.resolve(find(blocks, block => block.id === id))
}
