// app/api/mock/report.js
import {max, find} from "app/core/utils"
import resolve from "app/api/mock/resolve"

const reports = [{
  id: 1,
  name: "Demo 1",
  views: [
    {
      id: 1,
      type: "map",
      name: "Map 1",
      layers: [
        {
          type: "area",
          source: "US_COUNTY"
        },
        {
          type: "heat",
          source: "US_COUNTY"
        },
        {
          type: "bubble",
          source: "US_STATE"
        },
        {
          type: "dot",
          source: "US_COUNTY"
        },
        {
          type: "column",
          source: "US_STATE"
        }
      ]
    },
    {
      id: 2,
      type: "map",
      name: "Map 2",
      layers: [
        {
          type: "heat",
          source: "US_COUNTY"
        },
        {
          type: "bubble",
          source: "US_STATE"
        },
        {
          type: "column",
          source: "US_STATE"
        }
      ]
    }
  ]
}]

reports.forEach(report => {
  report.nextViewId = max(report.views, view => view.id).id + 1
})

let nextReportId = max(reports, report => report.id).id + 1

export function create(report) {
  report = {
    id: nextReportId++,
    ...report
  }
  reports.push(report)
  return Promise.resolve(report)
}

export function findById(id) {
  return resolve(find(reports, report => report.id === id))
}
