// app/api/mock/report.js
import {find} from "app/core/utils"

const reports = []
let nextReportId = 1

function createReport(options) {
  const report = {
    ...options,
    id: nextReportId++,
    nextViewId: 1,
    views: []
  }
  reports.push(report)
  return report
}

function addView(report, view) {
  view.id = report.nextViewId++
  report.views.push(view)
  return view
}

const report1 = createReport({
  name: "Demo"
})
addView(report1, {
  type: "map",
  name: "Map 1",
  layers: [
    {type: "area", source: "US_COUNTY"},
    {type: "heat", source: "US_COUNTY"},
    {type: "bubble", source: "US_STATE"},
    {type: "dot", source: "US_COUNTY"},
    {type: "column", source: "US_STATE"}
  ]
})
addView(report1, {
  type: "map",
  name: "Map 2",
  layers: [
    {type: "heat", source: "US_COUNTY"},
    {type: "bubble", source: "US_STATE"},
    {type: "column", source: "US_STATE"}
  ]
})

export function findAll() {
  return Promise.resolve(reports)
}

export function findById(id) {
  return Promise.resolve(find(reports, report => report.id === id))
}
