// app/block/_client/views/WidgetChartBar.js
import React from "react"
import {findDOMNode} from "react-dom"

const {Component, PropTypes} = React

class WidgetChartBar extends Component {

  static propTypes = {
    block: PropTypes.object.isRequired,
    config: PropTypes.object.isRequired,
    data: PropTypes.array.isRequired
  }

  componentDidMount() {
    this.updateChart()
  }

  componentDidUpdate() {
    this.updateChart()
  }

  updateChart() {
    const {block, config, data} = this.props
    const chartOptions = {
      title: false,
      credits: false,
      chart: {
        type: "bar"
      }
    }
    const categories = []
    for (let year = block.loader.year[0]; year <= block.loader.year[1]; ++year) {
      categories.push(year)
    }
    chartOptions.xAxis = {
      categories
    }
    chartOptions.yAxis = {
      title: {
        text: config.yAxis.label
      }
    }
    chartOptions.series = config.series.map(seriesConfig => ({
      name: seriesConfig.label,
      data: data.map(record => +record[seriesConfig.id].replace(/,/g, ""))
    }))
    const {$} = window
    $(findDOMNode(this)).highcharts(chartOptions)
  }

  render() {
    return <div></div>
  }

}

export default WidgetChartBar
