// app/_client/sampleBlocks.js
const sampleBlocks = [
  {
    id: 1,
    loader: {
      type: "usda/state",
      fields: [
        {
          id: "planted",
          source: "NASS",
          params: {
            sourceId: 2, sectorId: 2, groupId: 8, commodityId: 66,
            statisticCatId: 21, classId: 145, utilPracticeId: 2, prodnPracticeId: 14, unitId: 35,
            year: 2014
          }
        },
        {
          id: "harvested",
          source: "NASS",
          params: {
            sourceId: 2, sectorId: 2, groupId: 8, commodityId: 66,
            statisticCatId: 15, classId: 145, utilPracticeId: 34, prodnPracticeId: 14, unitId: 35,
            year: 2014
          }
        }
      ]
    },
    widget: {
      type: "table",
      columns: [
        {
          id: "name",
          label: "State"
        },
        {
          id: "planted",
          label: "Area planted for all purposes"
        },
        {
          id: "harvested",
          label: "Area harvested for grain"
        }
      ]
    }
  }
]

export default sampleBlocks
