// app/web/mock/reports.js
const reports = [
  {
    id: 1,
    name: "Demo 1",
    defaultView: 1,
    views: [
      {
        id: 1,
        name: "Map 1",
        type: "map",
        layers: [
          {
            type: "area",
            category: 1,
            locationType: "US_COUNTY"
          },
          {
            type: "heat",
            category: 2,
            locationType: "US_COUNTY"
          },
          {
            type: "dot",
            category: 3,
            locationType: "US_COUNTY"
          },
          {
            type: "bubble",
            category: 4,
            locationType: "US_STATE"
          },
          {
            type: "column",
            category: 5,
            locationType: "US_STATE"
          }
        ]
      },
      {
        id: 2,
        name: "Map 2",
        type: "map",
        layers: [
          {
            type: "heat",
            category: 5,
            locationType: "US_COUNTY"
          },
          {
            type: "bubble",
            category: 3,
            locationType: "US_STATE"
          },
          {
            type: "column",
            category: 4,
            locationType: "US_STATE"
          }
        ]
      },
      {
        id: 3,
        name: "Trends 1",
        type: "trends",
        timeRange: "last5",
        categories: [1, 2, 3],
        locationType: "US_STATE",
        locations: ["34", "37", "05", "41"]
      },
      {
        id: 4,
        name: "Trends 2",
        type: "trends",
        timeRange: "last10",
        categories: [4, 5],
        locationType: "US_STATE",
        locations: ["01", "02", "04", "05", "06", "08", "09", "10"]
      },
      {
        id: 5,
        name: "Trends 3",
        type: "trends",
        timeRange: "last10",
        categories: [1, 2, 3, 4, 5],
        locationType: "US_COUNTY",
        locations: ["20001", "20003", "20005", "20007", "20009", "20011", "20013", "20015", "20017", "20019", "20021"]
      }
    ]
  }
]

export default reports
