// app/web/mock/reports.js
const reports = [
  {
    id: 1,
    name: "Demo 1",
    defaultView: 1,
    views: [
      {
        id: 1,
        name: "Map 1",
        type: "map",
        layers: [
          {
            type: "area",
            category: 1,
            location: "US_COUNTY"
          },
          {
            type: "heat",
            category: 2,
            location: "US_COUNTY"
          },
          {
            type: "dot",
            category: 3,
            location: "US_COUNTY"
          },
          {
            type: "bubble",
            category: 4,
            location: "US_STATE"
          },
          {
            type: "column",
            category: 5,
            location: "US_STATE"
          }
        ]
      },
      {
        id: 2,
        name: "Map 2",
        type: "map",
        layers: [
          {
            type: "heat",
            category: 5,
            location: "US_COUNTY"
          },
          {
            type: "bubble",
            category: 3,
            location: "US_STATE"
          },
          {
            type: "column",
            category: 4,
            location: "US_STATE"
          }
        ]
      },
      {
        id: 3,
        name: "Trends 1",
        type: "trends"
      },
      {
        id: 4,
        name: "Trends 2",
        type: "trends"
      },
      {
        id: 5,
        name: "Trends 3",
        type: "trends"
      }
    ]
  }
]

export default reports
