// app/web/mock/reports.js
const reports = [
  {
    id: 1,
    name: "Demo 1",
    defaultView: 1,
    views: [
      {
        id: 1,
        type: "map",
        name: "Map 1",
        layers: [
          {
            type: "area",
            source: "US_COUNTY"
          },
          {
            type: "heat",
            source: "US_COUNTY"
          },
          {
            type: "bubble",
            source: "US_STATE"
          },
          {
            type: "dot",
            source: "US_COUNTY"
          },
          {
            type: "column",
            source: "US_STATE"
          }
        ]
      },
      {
        id: 2,
        type: "map",
        name: "Map 2",
        layers: [
          {
            type: "heat",
            source: "US_COUNTY"
          },
          {
            type: "bubble",
            source: "US_STATE"
          },
          {
            type: "column",
            source: "US_STATE"
          }
        ]
      },
      {
        id: 3,
        type: "list",
        name: "List 1"
      },
      {
        id: 4,
        type: "list",
        name: "List 2"
      },
      {
        id: 5,
        type: "trendsBar",
        name: "Trends 1"
      },
      {
        id: 6,
        type: "trendsLine",
        name: "Trends 2"
      },
      {
        id: 7,
        type: "trendsArea",
        name: "Trends 3"
      }
    ]
  }
]

export default reports
