// app/api/index.js
import {getReport, createReport, getTrendsData, getMapData} from "app/api/handlers/report"
import {getGeosBySource} from "app/api/handlers/geo"
import {createUser, login, logout} from "app/api/handlers/user"
import {getSchema as getSignupSchema} from "app/web/forms/signup"
import {getSchema as getLoginSchema} from "app/web/forms/login"
import {getSchema as getReportSchema} from "app/web/forms/report"
import {handlePromise} from "app/core/utils/handler"

const authRequired = {
  auth: {
    mode: "required"
  }
}

export function register(server, config, next) {
  server.route([
    {
      method: "POST",
      path: "/user",
      handler: handlePromise(createUser, getSignupSchema())
    },
    {
      method: "POST",
      path: "/login",
      handler: handlePromise(login, getLoginSchema())
    },
    {
      method: "GET",
      path: "/logout",
      handler: handlePromise(logout)
    },
    {
      method: "GET",
      path: "/geo/source/{source}",
      handler: handlePromise(getGeosBySource)
    },
    {
      method: "GET",
      path: "/report/{id}",
      handler: handlePromise(getReport)
    },
    {
      method: "POST",
      path: "/report",
      handler: handlePromise(createReport, getReportSchema()),
      config: authRequired
    },
    {
      method: "GET",
      path: "/report/{id}/view/{viewId}/trendsData",
      handler: handlePromise(getTrendsData)
    },
    {
      method: "GET",
      path: "/report/{id}/view/{viewId}/mapData",
      handler: handlePromise(getMapData)
    }
  ])
  next()
}

register.attributes = {
  name: "agridash-api"
}
