// app/web/proxy.js
import h2o2 from "h2o2"
import {startsWith} from "app/core/utils"

function createMapRoute(connection, method = "GET") {
  return {
    method,
    path: connection.url + "/{path*}",
    handler: {
      proxy: {
        passThrough: true,
        xforward: true,
        mapUri: (request, done) => {
          done(null, "http://" + connection.host + ":" + connection.port + "/" + request.params.path
            + (request.url.search || ""))
        }
      }
    }
  }
}

export function register(server, config, next) {
  const routes = []
  if (startsWith(config.api.url, "/")) {
    routes.push(createMapRoute(config.api))
  }
  if (startsWith(config.tile.url, "/")) {
    routes.push(createMapRoute(config.tile))
  }
  if (config.webpack.publicPath) {
    routes.push({
      method: "GET",
      path: config.webpack.publicPath + "{path*}",
      handler: {
        proxy: {
          host: config.webpack.host,
          port: config.webpack.port,
          passThrough: true,
          xforward: true
        }
      }
    })
  }
  if (routes.length) {
    server.register(h2o2, next)
    server.route(routes)
  }
}

register.attributes = {
  name: "agridash-proxy"
}
