// app/report/_server/handlers/sampleBlocks.js
const sampleBlocks = [
  {
    id: 1,
    title: "Corn, planted vs harvested",
    loader: {
      type: "usda/state",
      fields: [
        {
          id: "planted",
          source: "NASS",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 66,
            statisticcatid: 21, classid: 145, utilpracticeid: 2, prodnpracticeid: 14, unitid: 35,
            year: 2014
          }
        },
        {
          id: "harvested",
          source: "NASS",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 66,
            statisticcatid: 15, classid: 145, utilpracticeid: 34, prodnpracticeid: 14, unitid: 35,
            year: 2014
          }
        },
        {
          id: "yield",
          source: "NASS",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 66,
            statisticcatid: 179, classid: 145, utilpracticeid: 109, prodnpracticeid: 14, unitid: 238,
            year: 2014
          }
        }
      ]
    },
    widget: {
      type: "table",
      columns: [
        {
          id: "name",
          label: "State"
        },
        {
          id: "planted",
          label: "Area planted for all purposes"
        },
        {
          id: "harvested",
          label: "Area harvested for grain"
        },
        {
          id: "yield",
          label: "Yield"
        }
      ]
    }
  },
  {
    id: 2,
    title: "Soybeans, planted vs harvested",
    loader: {
      type: "usda/state",
      fields: [
        {
          id: "planted",
          source: "NASS",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 303,
            statisticcatid: 21, classid: 145, utilpracticeid: 2, prodnpracticeid: 14, unitid: 35,
            year: 2014
          }
        },
        {
          id: "harvested",
          source: "NASS",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 303,
            statisticcatid: 15, classid: 145, utilpracticeid: 2, prodnpracticeid: 14, unitid: 35,
            year: 2014
          }
        }
      ]
    },
    widget: {
      type: "table",
      columns: [
        {
          id: "name",
          label: "State"
        },
        {
          id: "planted",
          label: "Area planted"
        },
        {
          id: "harvested",
          label: "Area harvested"
        }
      ]
    }
  }
]

export default sampleBlocks
