// app/report/_server/handlers/sampleBlocks.js
const sampleBlocks = [
  {
    id: 1,
    title: "Corn, planted vs harvested, 2014 (table)",
    loader: {
      type: "groupByState",
      fields: [
        {
          id: "planted",
          type: "nass/getCommodityValue",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 66,
            statisticcatid: 21, classid: 145, utilpracticeid: 2, prodnpracticeid: 14, unitid: 35,
            year: 2014
          }
        },
        {
          id: "harvested",
          type: "nass/getCommodityValue",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 66,
            statisticcatid: 15, classid: 145, utilpracticeid: 34, prodnpracticeid: 14, unitid: 35,
            year: 2014
          }
        },
        {
          id: "yield",
          type: "nass/getCommodityValue",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 66,
            statisticcatid: 179, classid: 145, utilpracticeid: 109, prodnpracticeid: 14, unitid: 238,
            year: 2014
          }
        }
      ]
    },
    widget: {
      type: "table",
      columns: [
        {
          id: "name",
          label: "State"
        },
        {
          id: "planted",
          label: "Area planted for all purposes"
        },
        {
          id: "harvested",
          label: "Area harvested for grain"
        },
        {
          id: "yield",
          label: "Yield"
        }
      ]
    }
  },
  {
    id: 2,
    title: "Corn, planted vs harvested, COLORADO, 2000 to 2014 (line)",
    loader: {
      type: "groupByYear",
      year: [2000, 2014],
      fields: [
        {
          id: "planted",
          type: "nass/getCommodityValue",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 66,
            statisticcatid: 21, classid: 145, utilpracticeid: 2, prodnpracticeid: 14, unitid: 35,
            stateid: 8
          }
        },
        {
          id: "harvested",
          type: "nass/getCommodityValue",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 66,
            statisticcatid: 15, classid: 145, utilpracticeid: 34, prodnpracticeid: 14, unitid: 35,
            stateid: 8
          }
        }
      ]
    },
    widget: {
      type: "chartLine",
      yAxis: {
        label: "Area"
      },
      series: [
        {
          id: "planted",
          label: "Planted for all purposes"
        },
        {
          id: "harvested",
          label: "Harvested for grain"
        }
      ]
    }
  },
  {
    id: 3,
    title: "Corn, planted vs harvested, US TOTAL, 2000 to 2014 (column)",
    loader: {
      type: "groupByYear",
      year: [2000, 2014],
      fields: [
        {
          id: "planted",
          type: "nass/getCommodityValue",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 66,
            statisticcatid: 21, classid: 145, utilpracticeid: 2, prodnpracticeid: 14, unitid: 35,
            agglevelid: 4, stateid: 99
          }
        },
        {
          id: "harvested",
          type: "nass/getCommodityValue",
          params: {
            sourceid: 2, sectorid: 2, groupid: 8, commodityid: 66,
            statisticcatid: 15, classid: 145, utilpracticeid: 34, prodnpracticeid: 14, unitid: 35,
            agglevelid: 4, stateid: 99
          }
        }
      ]
    },
    widget: {
      type: "chartColumn",
      yAxis: {
        label: "Area"
      },
      series: [
        {
          id: "planted",
          label: "Planted for all purposes"
        },
        {
          id: "harvested",
          label: "Harvested for grain"
        }
      ]
    }
  }
]

export default sampleBlocks
