// app/web/views/LoginForm.js
import React from "react"
import {connect} from "react-redux"
import createForm from "app/web/forms/form"
import {getSchema as getLoginSchema, getOptions as getLoginOptions} from "app/web/forms/login"
import {updateInfo} from "app/web/actions/info"

const {Component, PropTypes} = React

class LoginForm extends Component {

  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
    location: PropTypes.object.isRequired
  }

  componentDidMount() {
    this.removeForm = createForm(this.refs.container, getLoginSchema(), getLoginOptions(), {
      url: "/login",
      buttonLabel: "Log In",
      done: user => {
        const {location} = this.props
        const nextPath = (location.state && location.state.nextPath) || `/user/${user.username}`
        this.props.dispatch(updateInfo({
          user
        }))
        this.props.history.replaceState(null, nextPath)
      }
    })
  }

  componentWillUnmount() {
    if (this.removeForm) {
      this.removeForm()
    }
  }

  render() {
    return <div ref="container"></div>
  }

}

export default connect()(LoginForm)
