// app/web/views/MapView.js
import L from "leaflet"
import React from "react"
import {container} from "app/web/views/MapView.css"

const {Component, PropTypes} = React

const combine = true

class MapView extends Component {

  static propTypes = {
    info: PropTypes.object.isRequired,
    report: PropTypes.object.isRequired,
    view: PropTypes.object.isRequired,
    resultId: PropTypes.number.isRequired
  }

  componentDidMount() {
    const {info, report, view, resultId} = this.props
    const {tileUrl, sessionId} = info
    this.map = L.map(this.refs.container, {
      attributionControl: false,
      fullscreenControl: true,
      zoomAnimation: false,
      center: [37.8, -96.9],
      zoom: 4
    })
    // L.tileLayer("https://{s}.tiles.mapbox.com/v4/{style}/{z}/{x}/{y}.png?access_token={key}", {
    //   style: "mapbox.light",
    //   key: "pk.eyJ1IjoiYWdyaWRhc2giLCJhIjoiYjdiZjc4YTA0ZjhhODg0N2VkMDM0ODQwNGUwNDA0ZDQifQ.jmfdrMQc252iW2ovBPRTOg"
    // }).addTo(this.map)
    if (combine) {
      this.tileLayer = L.tileLayer(`${tileUrl}/report/${report.id}/view/${view.id}/{z}/{x}/{y}.png`
        + `?sessionId=${sessionId}&resultId=${resultId}`).addTo(this.map)
    } else {
      L.tileLayer(`${tileUrl}/report/${report.id}/view/${view.id}/{z}/{x}/{y}.png?h=1,2,3,4`
        + `&sessionId=${sessionId}&resultId=${resultId}`).addTo(this.map)
      L.tileLayer(`${tileUrl}/report/${report.id}/view/${view.id}/{z}/{x}/{y}.png?h=0,2,3,4`
        + `&sessionId=${sessionId}&resultId=${resultId}`).addTo(this.map)
      L.tileLayer(`${tileUrl}/report/${report.id}/view/${view.id}/{z}/{x}/{y}.png?h=0,1,3,4`
        + `&sessionId=${sessionId}&resultId=${resultId}`).addTo(this.map)
      L.tileLayer(`${tileUrl}/report/${report.id}/view/${view.id}/{z}/{x}/{y}.png?h=0,1,2,4`
        + `&sessionId=${sessionId}&resultId=${resultId}`).addTo(this.map)
      L.tileLayer(`${tileUrl}/report/${report.id}/view/${view.id}/{z}/{x}/{y}.png?h=0,1,2,3`
        + `&sessionId=${sessionId}&resultId=${resultId}`).addTo(this.map)
    }
  }

  componentDidUpdate() {
    if (!combine) {
      return
    }
    const {info, report, view, resultId} = this.props
    const {tileUrl, sessionId} = info
    this.tileLayer.setUrl(`${tileUrl}/report/${report.id}/view/${view.id}/{z}/{x}/{y}.png`
      + `?sessionId=${sessionId}&resultId=${resultId}`)
  }

  componentWillUnmount() {
    this.map.remove()
    this.map = null
    this.tileLayer = null
  }

  render() {
    return <div ref="container" className={container}>
    </div>
  }

}

export default MapView
