// app/web/views/MapView.js
import React from "react"
import {findDOMNode} from "react-dom"
import styles from "app/web/views/MapView.css"

const {Component, PropTypes} = React

class MapView extends Component {

  static propTypes = {
    report: PropTypes.object.isRequired,
    view: PropTypes.object.isRequired
  }

  componentDidMount() {
    const {L, clientConfig} = window
    const {report, view} = this.props
    this.map = L.map(findDOMNode(this), {
      attributionControl: false,
      fullscreenControl: true,
      zoomAnimation: false,
      center: [37.8, -96.9],
      zoom: 4
    })
    // L.tileLayer("https://{s}.tiles.mapbox.com/v4/{style}/{z}/{x}/{y}.png?access_token={key}", {
    //   style: "mapbox.light",
    //   key: "pk.eyJ1IjoiYWdyaWRhc2giLCJhIjoiYjdiZjc4YTA0ZjhhODg0N2VkMDM0ODQwNGUwNDA0ZDQifQ.jmfdrMQc252iW2ovBPRTOg"
    // }).addTo(this.map)
    const combine = true
    if (combine) {
      this.tileLayer = L.tileLayer(`${clientConfig.tileUrl}/${report.id}/${view.id}/{z}/{x}/{y}.png`).addTo(this.map)
    } else {
      L.tileLayer(`${clientConfig.tileUrl}/${report.id}/${view.id}/{z}/{x}/{y}.png?h=1,2,3,4`).addTo(this.map)
      L.tileLayer(`${clientConfig.tileUrl}/${report.id}/${view.id}/{z}/{x}/{y}.png?h=0,2,3,4`).addTo(this.map)
      L.tileLayer(`${clientConfig.tileUrl}/${report.id}/${view.id}/{z}/{x}/{y}.png?h=0,1,3,4`).addTo(this.map)
      L.tileLayer(`${clientConfig.tileUrl}/${report.id}/${view.id}/{z}/{x}/{y}.png?h=0,1,2,4`).addTo(this.map)
      L.tileLayer(`${clientConfig.tileUrl}/${report.id}/${view.id}/{z}/{x}/{y}.png?h=0,1,2,3`).addTo(this.map)
    }
  }

  componentDidUpdate() {
    const {clientConfig} = window
    const {report, view} = this.props
    this.tileLayer.setUrl(`${clientConfig.tileUrl}/${report.id}/${view.id}/{z}/{x}/{y}.png`)
  }

  componentWillUnmount() {
    this.map.remove()
    this.map = null
    this.tileLayer = null
  }

  render() {
    return <div className={styles.container}>
    </div>
  }

}

export default MapView
