// app/web/views/Report.js
import React from "react"
import cx from "classnames"
import styles from "app/web/views/Report.css"
import Map from "app/web/views/Map"
import NavLink from "app/web/views/NavLink"
import Icon from "app/web/views/Icon"
import Button from "app/web/views/Button"
import {find} from "app/core/utils"
import reports from "app/web/mock/reports"

const {Component, PropTypes} = React

const viewByType = {
  map: {
    icon: "map-o",
    component: Map
  },
  list: {
    icon: "list-ul"
  },
  trendsBar: {
    icon: "bar-chart"
  },
  trendsLine: {
    icon: "line-chart"
  },
  trendsArea: {
    icon: "area-chart"
  }
}

class Report extends Component {

  static propTypes = {
    params: PropTypes.object.isRequired,
    location: PropTypes.object.isRequired
  }

  render() {
    const {params, location} = this.props
    const report = find(reports, "id", +params.id)
    const activeViewId = +location.query.view
    const activeView = find(report.views, "id", activeViewId)
    const ViewComponent = viewByType[activeView.type].component
    return <div>
      <div className="row">
        <div className="col-sm-4 col-md-3">
          <div className="text-center m-y-md">
            <h3>{report.name}</h3>
            <div className="btn-group btn-group-sm">
              <Button className="btn-secondary"><Icon icon="thumbs-o-up"/></Button>
              <Button className="btn-secondary"><Icon icon="link"/></Button>
              <Button className="btn-secondary"><Icon icon="clone"/></Button>
            </div>
          </div>
          <ul className="nav nav-pills nav-stacked m-b">
            {/* Must use location.pathname here rackt/react-router/issues/1828 */}
            {report.views.map(view => <NavLink key={view.id} to={location.pathname} query={{view: view.id}}
                className={cx(view.id === activeViewId && "active")}>
              <Icon icon={viewByType[view.type].icon} fw/> {view.name}
            </NavLink>)}
          </ul>
        </div>
        <div className="col-sm-8 col-md-9">
          <div className={styles.view}>
            <ViewComponent view={activeView}/>
          </div>
        </div>
      </div>
    </div>
  }

}

export default Report
