// app/web/views/TrendsView.js
import React from "react"
import {findDOMNode} from "react-dom"
import {keys} from "app/core/utils"
import styles from "app/web/views/TrendsView.css"

const {Component, PropTypes} = React

class TrendsView extends Component {

  static propTypes = {
    data: PropTypes.array.isRequired
  }

  componentDidMount() {
    const {$} = window
    const {data} = this.props
    const years = keys(data[0].data)
    years.sort()
    const element = $(findDOMNode(this))
    element.highcharts({
      title: false,
      credits: false,
      xAxis: {
        title: false,
        categories: years
      },
      yAxis: {
        title: false
      },
      legend: {
        maxHeight: element.height() / 5
      },
      plotOptions: {
        series: {
          animation: {
            duration: 250
          }
        }
      },
      series: data.map(record => ({
        name: `${record.name} - ${record.categoryName}`,
        data: years.map(year => record.data[year])
      }))
    })
  }

  render() {
    return <div className={styles.container}>
    </div>
  }

}

export default TrendsView
